src/Dependencies/Collections/SegmentedArray.cs

```diff
@@ -151,14 +151,14 @@ public static void Copy<T>(SegmentedArray<T> sourceArray, int sourceIndex, Array
                 throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
 
             var copied = 0;
-            foreach (var memory in sourceArray.GetSegments(0, length))
+            foreach (var memory in sourceArray.GetSegments(sourceIndex, length))
             {
                 if (!MemoryMarshal.TryGetArray<T>(memory, out var segment))
                 {
                     throw new NotSupportedException();
                 }
 
-                Array.Copy(segment.Array!, sourceIndex: segment.Offset, destinationArray: destinationArray, destinationIndex: copied, length: segment.Count);
+                Array.Copy(segment.Array!, sourceIndex: segment.Offset, destinationArray: destinationArray, destinationIndex: destinationIndex + copied, length: segment.Count);
```

Comment:
I might not understand this properly—once we reach the end of what we're copying, should the length still be segment.Count or should it be the number of Ts remaining?

Reply:
➡️ `memory` is a `Memory<T>, and `segment` is an `ArraySegment<T>` representing the slice of a `T[]` captured by `memory`. `GetSegments` ensures that the complete sequence contains exactly `length` items.